samples/s3-cp/s3-cp.c

```diff
@@ -0,0 +1,403 @@
+    fprintf(stderr, " destination_bucket: the S3 bucket the object will be copied to\n");
+    fprintf(stderr, " destination_object_key: the key to be used for the new S3 object\n");
+    fprintf(stderr, "\n Options:\n\n");
+    fprintf(stderr, "  -r, --region: the AWS region where the bucket is located. Default is us-west-2.\n");
+    fprintf(stderr, "  -v, --verbose: ERROR|INFO|DEBUG|TRACE: log level to configure. Default is none.\n");
+    fprintf(stderr, "  -h, --help\n");
+    fprintf(stderr, "            Display this message and quit.\n");
+    exit(exit_code);
+}
+
+static struct aws_cli_option s_long_options[] = {
+    {"region", AWS_CLI_OPTIONS_REQUIRED_ARGUMENT, NULL, 'r'},
+    {"verbose", AWS_CLI_OPTIONS_REQUIRED_ARGUMENT, NULL, 'v'},
+    {"help", AWS_CLI_OPTIONS_NO_ARGUMENT, NULL, 'h'},
+    /* Per getopt(3) the last element of the array has to be filled with all zeros */
+    {NULL, AWS_CLI_OPTIONS_NO_ARGUMENT, NULL, 0},
+};
+
+static void s_parse_options(int argc, char **argv, struct app_ctx *ctx) {
+    while (true) {
+        int option_index = 0;
+        int c = aws_cli_getopt_long(argc, argv, "r:v:h", s_long_options, &option_index);
+        if (c == -1) {
+            break;
+        }
+
+        switch (c) {
+            case 0:
+                /* getopt_long() returns 0 if an option.flag is non-null */
+                break;
+            case 'r':
+                ctx->region = aws_cli_optarg;
+                break;
+            case 'v':
+                if (!strcmp(aws_cli_optarg, "TRACE")) {
+                    ctx->log_level = AWS_LL_TRACE;
+                } else if (!strcmp(aws_cli_optarg, "INFO")) {
+                    ctx->log_level = AWS_LL_INFO;
+                } else if (!strcmp(aws_cli_optarg, "DEBUG")) {
+                    ctx->log_level = AWS_LL_DEBUG;
+                } else if (!strcmp(aws_cli_optarg, "ERROR")) {
+                    ctx->log_level = AWS_LL_ERROR;
+                } else {
+                    fprintf(stderr, "unsupported log level %s.\n", aws_cli_optarg);
+                    s_usage(1);
+                }
+                break;
+            case 'h':
+                s_usage(0);
+                break;
+            default:
+                fprintf(stderr, "Unknown option\n");
+                s_usage(1);
+        }
+    }
+
+    if (aws_cli_optind < argc) {
+        struct aws_byte_cursor source_uri_cursor = aws_byte_cursor_from_c_str(argv[aws_cli_optind++]);
+
+        if (aws_uri_init_parse(&ctx->source_uri, ctx->allocator, &source_uri_cursor)) {
+            fprintf(
+                stderr,
+                "Failed to parse uri %s with error %s\n",
+                (char *)source_uri_cursor.ptr,
+                aws_error_debug_str(aws_last_error()));
+            s_usage(1);
+        };
+    } else {
+        fprintf(stderr, "An URI for the source object must be supplied.\n");
+        s_usage(1);
+    }
+
+    if (aws_cli_optind < argc) {
```

Comment:
don't you need to check that both of these branches were taken?

Reply:
It is currently expecting the last two parameters to be the source and target URIs respectively. If any is missing, it will print the corresponding error.